### PR TITLE
feat: a new install_code mode

### PR DESCRIPTION
**Pull request by roman-kashitsyn**
This PR describes a new variant of the upgrade mode for `install_code` that allows canister controllers to skip the `pre_upgrade` hook during the upgrade.

This upgrade method can serve two purposes.

  1. Act as an escape hatch if the canister entered otherwise irrecoverable state due to a critical bug in the `post_upgrade` method of the new canister module. The stable memory must already contain the entire state for the "fix" to work.

  2. Canister developers can use it during canister deployment to excercise the `pre_upgrade` method of the new canister.

Recovering from buggy `post_upgrade` hooks
-------------

Imagine that the `post_upgrade` method implementation has a bug that bricks the canister. Upgrading the canister in such a state is either impossible (`pre_upgrade` method traps unconditionally) or can result in a data loss (the `post_upgrade` method erroneously discarded important data, and executing `pre_upgrade` would corrupt the original).  The stable memory, however, might still contain the correct state that the old module recorded.

If developers ever find themselves in such situation (I did), the `upgrade` install method with `skip_pre_upgrade` is the only salvation.

Testing the `pre_upgrade` hook
--------------

Testing `pre_upgrade` hooks is a constant source of struggle. If this method contains a bug, the canister can become unupgradable.

The upgrade method with `skip_pre_upgrade` provides a method to excercise the `pre_upgrade` method of the new canister module without the fear of data loss using the following procedure:

  1. Stop the canister.
  2. Upgrade the canister to the new module.
  3. If upgrade succeeded, upgrade the canister again to the same module. This upgrade will excercise the `pre_upgrade` hook.
  4. If the second upgrade succeeds, proceed with the next step. Otherwise, install the old canister module in upgrade mode with `skip_pre_upgrade`.
  5. Start the canister.

### CODE REVIEW

File: spec/index.adoc
Comment (mraszyk) [inline spec/index.adoc:988]: canister version is now the integer bumped upon changes to the canister (e.g., upgrade or change of controllers); I'd refer here to "canister module and state" instead of "canister version"

File: spec/index.adoc
Comment (mraszyk) [inline spec/index.adoc:3498]: This mode should also be introduced in the type definition in the section "The system state"

Comment (mraszyk) [reply spec/index.adoc:3498]: Nevermind; this comment affects the canister history MR after this one is merged.

**Review by Dfinity-Bjoern — COMMENTED** [review]
Discussed in canister OS meeting, and generally approve of the change. We prefer to leave it as a open PR so it gets to master around the same time it gets to code and formal model.

**Comment by Dfinity-Bjoern** [comment]
This is currently still blocked by the Candid bug that prevents backward-compatible decoding of extended variants.

File: spec/ic.did
Comment (dsarlis) [reply spec/ic.did:93]: @Dfinity-Bjoern @roman-kashitsyn wdyt?

Comment (ielashi) [reply spec/ic.did:93]: Ping? :)

Comment (Dfinity-Bjoern) [reply spec/ic.did:93]: May also make sense whether the bug has been fixed in the meantime? @chenyan-dfinity

Comment (chenyan-dfinity) [reply spec/ic.did:93]: Both extending `upgrade` to `upgrade : opt record` and adding a new variant tag in the function argument are non-breaking changes. This is always supported and doesn't rely on the extensible variant fix.

Comment (mraszyk) [reply spec/ic.did:93]: > doesn't rely on the extensible variant fix

I believe the question was whether the "extensible variant" issue has been fixed in the meantime.

Comment (dsarlis) [reply spec/ic.did:93]: @chenyan-dfinity The problem is not extending a function argument but the fact that these install code messages might need to fly between subnets and therefore we would also need to decode them on the receiving subnet. This is the problem with extending the variant tag and not that we'd break the interface of the management canister.

Comment (chenyan-dfinity) [reply spec/ic.did:93]: I don't see the difference. At the Candid level, everything proposed here is fine, it wouldn't break any existing clients, including the receiving subnet who is running an old client. This is an allowed change even without applying the extensible variant fix, which means it's working in production without the need to bump to candid 0.9.

Comment (mraszyk) [reply spec/ic.did:93]: Out of curiosity, I've run the following code:
```
use candid::{CandidType, Decode, Deserialize, Encode};
use serde::Serialize;

#[derive(CandidType, Deserialize, Serialize, Debug)]
pub enum CanisterInstallMode {
    #[serde(rename = "install")]
    Install = 1,
    #[serde(rename = "reinstall")]
    Reinstall = 2,
    #[serde(rename = "upgrade")]
    Upgrade = 3,
}

#[derive(CandidType, Deserialize, Serialize, Debug)]
pub enum CanisterInstallModeNew {
    #[serde(rename = "install")]
    Install = 1,
    #[serde(rename = "reinstall")]
    Reinstall = 2,
    #[serde(rename = "upgrade")]
    Upgrade = 3,
    #[serde(rename = "evict")]
    Evict = 4,
}

fn main() {
    let bytes = Encode!(&CanisterInstallMode::Install).unwrap();
    let mode_install = Decode!(&bytes, CanisterInstallModeNew).unwrap();
    let bytes = Encode!(&CanisterInstallMode::Reinstall).unwrap();
    let mode_reinstall = Decode!(&bytes, CanisterInstallModeNew).unwrap();
    let bytes = Encode!(&CanisterInstallMode::Upgrade).unwrap();
    let mode_upgrade = Decode!(&bytes, CanisterInstallModeNew).unwrap();
    println!(
        "install: {:?}; reinstall: {:?}; upgrade: {:?}",
        mode_install, mode_reinstall, mode_upgrade
    );
}
```
with the following `Cargo.toml`:
```
[package]
name = "subtyping"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
candid = "0.8"
ic-cdk = "0.7"
serde = "1"
```
and it prints
```
install: Install; reinstall: Reinstall; upgrade: Upgrade
```
as expected => the decoding seems to work fine.

Comment (chenyan-dfinity) [reply spec/ic.did:93]: Yep, `ModeNew` can also be changed to 
```
#[derive(CandidType, Deserialize, Debug)]
pub struct Flag { skip: bool }

#[derive(CandidType, Deserialize, Debug)]
pub enum CanisterInstallModeNew {
    #[serde(rename = "install")]
    Install,
    #[serde(rename = "reinstall")]
    Reinstall,
    #[serde(rename = "upgrade")]
    Upgrade(Option<Flag>),
    #[serde(rename = "evict")]
    Evict,
}
```
and still decodes fine.

Comment (dsarlis) [reply spec/ic.did:93]: @mraszyk You did not try to decode with the new variant "evict" using the old enum. That would not work. The problem is decoding an encoded blob containing the new variant while you're running the old code that doesn't contain it. I should've been more clear with my example of the two subnets communicating.

Here's the slack [thread](https://dfinity.slack.com/archives/C01DB8MQ5M1/p1675881598055839?thread_ts=1675868384.037619&cid=C01DB8MQ5M1) where I had attempted to add a new variant to `CanisterInstallMode` and we got an error because a subnet was sending an `install_code` with the new variant to another one that had the old version (without the new variant).

To confirm, use this modified main from your example:
```

Comment (dragoljub-duric) [reply spec/ic.did:93]: Continuing on @dsarlis message from above, we have two possible solutions:
1. Adding the new version of the `CanisterInstallMode` that would be the same as the old one while having the `Evict` mode, and the replica will always try to decode the newer version and if it fails, a replica will try to decode as the old version. The good thing about this approach is that code will be clean, and later we can remove the old version of the `CanisterInstallMode`. The bad thing is that this approach requires significant overhead because the `CanisterInstallMode` appears in a lot of the places in the replica and `NNS` canisters. Hence the plan will be, in the first release we should add a new version of `CanisterInstallMode`, in the second we should update the `NNS` canisters and start using `Evict` mode, and in the third release, we should remove the old `CanisterInstallMode`.
2. The other option is to add a flag to the `InstallCodeArgs` e.g. `skip_pre_upgrade`. The bad thing about this approach is that we will have the flag which exists for all the modes, but it applies only to one specific mode (`Upgrade`). The good thing is that this approach should be more straightforward, and easy to implement.
Wdyt @mraszyk @Dfinity-Bjoern ?

Comment (ielashi) [reply spec/ic.did:93]: > The bad thing about this approach is that we will have the flag which exists for all the modes, but it applies only to one specific mode (Upgrade).

How does this flag apply to all modes? It's only the property of the `upgrade` variant, so it only exists/applies to that mode.

Comment (chenyan-dfinity) [reply spec/ic.did:93]: I see. In this case, adding `evict` needs to be done in two steps: 1) Upgrade `mode` type to `opt variant` from `variant`; 2) Use Candid 0.9, and add the new `evict` tag to `mode`.

Another more hacky but easier solution is @dragoljub-duric's option 2. Upgrade `install_code` to be of type `(record {...}, opt bool)`, while the last `opt bool` only applies to `Mode::Upgrade`.

Comment (chenyan-dfinity) [reply spec/ic.did:93]: > The problem is decoding an encoded blob containing the new variant while you're running the old code that doesn't contain it.

@dsarlis This is a use case we never thought about. For canisters, we assume the upgrade is always done in one atomic step. Can we delay the client code upgrade until all subnets have caught up with the new replica? Otherwise, we essentially require both forward and backward compatibility, which is quite cumbersome to work with.

Comment (dsarlis) [reply spec/ic.did:93]: > This is a use case we never thought about. For canisters, we assume the upgrade is always done in one atomic step.

Tbh, I think this is a miss on Candid's part. In a distributed setting, having an IDL that supports backward and forward compatibility is a must imo. We can't/don't control all clients or servers and when they update, so it's very useful when you get these properties on the IDL level.

> Otherwise, we essentially require both forward and backward compatibility, which is quite cumbersome to work with.

I would actually say quite the opposite, that not having a way to easily get forward compatibility with Candid is a big PITA. This is in theory a trivial change ("add a new variant to an enum") and we need to bend our backs to do it in a way that does not break clients in a guaranteed way.

> If this feature is only rolled out to one subnet, I think it's okay to error out for other subnets?

Erroring out might be ok in some circumstances but not in others. If there's anything that starts failing because an error was returned during decoding which was somehow not expected, then that would be bad. It's hard for us to vouch that all usages are properly handling errors returned, so I'm leaning towards an approach that gives us the safety even if it means some more work/extra steps that need to happen. Add on top of this that we have an awkward dependency of the NNS canisters on the types used by the replica's management canister (that we should eventually remove) and the whole thing becomes even more convoluted.

Comment (chenyan-dfinity) [reply spec/ic.did:93]: > Add on top of this that we have an awkward dependency of the NNS canisters on the types used by the replica's management canister (that we should eventually remove) and the whole thing becomes even more convoluted.

Okay, that's the real problem. NNS is a client of the management canister. It should use the types generated from the deployed interface. Sharing types between service and client is making things unnecessarily difficult. In fact, I would say NNS canister should not be in the same repo as the replica to make testing easier.

In the real world, nobody upgrades the client before the new API is ready on the server side. It's expected for the old client to continue to work, while the new API is rolling out on the server side. But it's not expected for the server to respond to a future change that it is not aware of. Such system doesn't make sense mathematically.

Comment (dsarlis) [reply spec/ic.did]: > Okay, that's the real problem. NNS is a client of the management canister. It should use the types generated from the deployed interface. Sharing types between service and client is making things unnecessarily difficult. In fact, I would say NNS canister should not be in the same repo as the replica to make testing easier.

Yes, indeed, that's a problem as I said and I've personally been pushing the NNS team to change this (as part of another cleanup work that I believe is super important, specifically to use the external Rust cdk instead of the internal one). As for being in a separate repo -- I have no skin in that game, therefore no strong opinion.

> In the real world, nobody upgrades the client before the new API is ready on the server side. It's expected for the old client to continue to work, while the new API is rolling out on the server side. But it's not expected for the server to respond to a future change that it is not aware of.

Mostly agree except for the fact that you might have "peer-to-peer" architectures instead of true client-server. The case of replicas talking to each other is one such case but there are more in the real world. In that case, there's no "server updates before client" because basically the same process can be client _and_ server depending on the context. This is why protobuf, thrift, etc also handle forward compatibility because in a distributed system the concepts of client and server are some times blended.

> Such system doesn't make sense mathematically.

It does engineeringly (it's an actual world, I googled it) :)

Comment (dsarlis) [reply spec/ic.did]: I am ok with both, leaning towards adding the optional bool to upgrade, especially given that there's basically no difference in how we should handle the changes to avoid breakages.

Comment (mraszyk) [reply spec/ic.did]: I'm also fine with both extending the enum and turning the upgrade variant into a variant with an attached record. I wouldn't extend the arguments of install_code.

What hasn't been explicitly discussed though is how we deal with None values of Option<CanisterInstallMode> that we can get in step 2 of Yan Chen's proposal because it seems we're stuck with Option<CanisterInstallMode> forever then, right?

I didn't get the sentence "the replica will always try to decode the newer version and if it fails, a replica will try to decode as the old version"; in particular, what is the exact new version and old version here.

Comment (mraszyk) [reply spec/ic.did]: > when receiving they will first try to decode it as CanisterInstallMode_V2 and if it fails they will try to decode as CanisterInstallMode

in both cases a CanisterInstallMode can be decoded as CanisterInstallMode_V2 so decoding CanisterInstallMode_V2 can never fail => we don't need phase 3 and can already drop CanisterInstallMode in step 2

Comment (roman-kashitsyn) [reply spec/ic.did]: I like the `skip_pre_upgrade` idea, but I'd make the field optional as well (just in case):

```
      upgrade : opt record {
        skip_pre_upgrade: opt bool;
      }
```

Comment (Dfinity-Bjoern) [reply spec/ic.did]: So just to make sure we're all on the same page: the current suggestion is to essentially follow @ielashi's proposal above (https://github.com/dfinity/interface-spec/pull/135#discussion_r1212896298) but with @roman-kashitsyn's tweak making `skip_pre_upgrade` an `opt bool`? That seems fine by me.

File: spec/index.md
Comment (mraszyk) [inline spec/index.md:1101]:
```suggestion
The IC executes canister upgrade skip pre_upgrade as follows:
```

File: spec/index.md
Comment (ielashi) [inline spec/index.md:1091]: Now that it's a flag, I don't think we need this as a separate section. I'd just consolidate everything into a simple note.

```
:::note
The `skip_pre_upgrade` flag can be enabled to skip the execution of the`canister_pre_upgrade` method on the old canister instance.
The main purpose of this mode is recovery from cases when the `canister_pre_upgrade` hook traps unconditionally preventing the normal upgrade path.

Skipping the pre-upgrade can lead to data loss.
Use it only as the last resort and only if the stable memory already contains the entire canister state.
:::
```

Comment (dragoljub-duric) [reply spec/index.md]: Done, PTAL

File: spec/index.md
Comment (dragoljub-duric) [reply spec/index.md]: Done

File: spec/index.md
Comment (mraszyk) [inline spec/index.md:1093]:
```suggestion
The `skip_pre_upgrade` flag can be enabled to skip the execution of the `canister_pre_upgrade` method on the old canister instance.
```

File: spec/index.md
Comment (mraszyk) [inline spec/index.md:1092]: I wonder if we need a note here or if we should just make it part of the main text?

Comment (dragoljub-duric) [reply spec/index.md:1092]: I am fine with both, @ielashi wdyt?

Comment (ielashi) [reply spec/index.md:1092]: No strong opinion. Either is fine by me.

File: spec/ic.did
Comment (mraszyk) [reply spec/ic.did]: I've changed the type of the field `skip_pre_upgrade` from `bool` to `opt bool` and thus resolve this thread.

File: spec/index.md
Comment (mraszyk) [reply spec/index.md]: Kept the note, but simplified the surrounding text.

File: spec/index.md
Comment (crusso) [inline spec/index.md:1105]:
```suggestion
1.  Invokes `canister_pre_upgrade` (if exported by the current canister code and  `skip_pre_upgrade` is not `opt true` ) on the old instance, to give the canister a chance to clean up (e.g. move data to [stable memory](#system-api-stable-memory)).
```

File: spec/index.md
Comment (crusso) [inline spec/index.md:1837]:
```suggestion
-   If `mode =  variant { upgrade }`,  `mode = variant  { upgrade  = opt {skip_pre_upgrade = null} }`, or `mode = variant { upgrade = opt {skip_pre_upgrade = opt false} }`, this will perform an upgrade of a non-empty canister as described in [Canister upgrades](#system-api-upgrades), passing `arg` to the `canister_post_upgrade` method of the new instance.
```
The previous two bullets are also imprecise acc to candid syntax. I expect. I'm not even sure I'm getting it right.

File: spec/index.md
Comment (crusso) [inline spec/index.md:1839]:
```suggestion
-   If `mode = variant { upgrade = opt { record { skip_pre_upgrade = opt true} }`, the system handles this method similarly to the `mode = variant { upgrade }` case, except that it does not execute the `canister_pre_upgrade` method on the old instance.
```

**Review by crusso — COMMENTED** [review]
I really think it would be cleaner and easier to specify and understand to just add a new variant rather than extend update.

How about `update`, `replace` or `recover` or, indeed as originally proposed, `evict` ?

That would also make it easier to consume the interface for languages that support variant subtyping, but not the Candid optional field nonsense.

But mostly, it would make it easier to understand and write the spec.

But looking at all the previous comments, I guess I'm the odd one out. I won't die on this particular hill.

**Comment by crusso** [comment]
> ## Testing the `pre_upgrade` hook
> Testing `pre_upgrade` hooks is a constant source of struggle. If this method contains a bug, the canister can become unupgradable.
> 
> The upgrade method with `skip_pre_upgrade` provides a method to excercise the `pre_upgrade` method of the new canister module without the fear of data loss using the following procedure:
> 
> 1. Stop the canister.
> 2. Upgrade the canister to the new module.
> 3. If upgrade succeeded, upgrade the canister again to the same module. This upgrade will excercise the `pre_upgrade` hook.
> 4. If the second upgrade succeeds, proceed with the next step. Otherwise, install the old canister module in upgrade mode with `skip_pre_upgrade`.
> 5. Start the canister.

@roman-kashitsyn that might work, but isn't it simpler to do a trial upgrade to a canister that deliberately traps in post_upgrade? Then you go inspect response to see if you got the deliberate trap, or some other trap, indicating that something went wrong.That doesn't address the issue of how to fix a broken pre_upgrade (which this proposal might) but does let you test it non-destructively.

Cut'n'pasting a recipe I wrote up some time ago:

Testing upgrades

A developer might want to test an upgrade will complete successfully without losing state before performing the actual upgrade. One way to do this is to arrange for the canister to do a dry-run of the upgrade, that is guaranteed to fail with a trap in post_upgrade. After verifying the dry-run could complete successfully, the developer can perform the proper upgrade, this time without trapping, to commit to the change.

One way to code such a dry-run is to ensure that the new canister traps, either explicitly as the last step of canister/actor initialization or in a dedicated postupgrade Motoko system function. For added assurance, the canister could verify the restored state before trapping. By employing library function Debug.trap(msg), the developer can return a diagnostic error message, msg, along with the trap, to, for example, report the condition of the state post upgrade, as visible from the end of actor initialization or, optionally, within system function postupgrade.

Here’s a simple example of Counter class that can verify its own upgrade before committing to it:

https://m7sm4-2iaaa-aaaab-qabra-cai.ic0.app/?tag=2623570777

```
import Debug "mo:base/Debug";

actor class Counter(upgradeMode: {#verify; #commit}) {

    stable var version = 0;
    stable var state : Int = 0;

    func validate() : Bool {
        state >= 0;
    };
    public func inc() : async Int {
        state += 1;
        return state;
    };
    public query func read() : async {version : Nat; state: Int} {
        {version; state};
    };
    system func postupgrade() {
        version += 1;
        if (upgradeMode == #verify) {
          Debug.trap(debug_show 
            #verify {version; state; valid = validate()});
        };
    }
}
```

File: spec/index.md
Comment (mraszyk) [reply spec/index.md:1837]: > The previous two bullets (for mode = install and mode = reinstall` are also imprecise acc to candid syntax, but I can't suggest on those line.

Fixed.

Comment (mraszyk) [reply spec/index.md:1837]: Your suggestion LGTM.

File: spec/index.md
Comment (mraszyk) [inline spec/index.md:3689]:
```suggestion
Only the controllers of the given canister can install new code. This changes the code of an *existing* canister, preserving the state in the stable memory. This involves invoking the `canister_pre_upgrade` method, if the `skip_pre_upgrade` flag is not set to `opt true`, on the old and `canister_post_upgrade` method on the new canister, which must succeed and must not invoke other methods.
```

**Comment by oggy-dfin** [comment]
I think this is an improvement over the status quo so we should go through with this (if we were doing it from the scratch I'd advocate removing this whole state management business from the System API and pushing into a "user space" library, but that ship's long sailed). I don't have a strong opinion on evict vs. upgrade with flags. From a UX perspective I dislike `skip_pre_upgrade` being an `opt bool` instead of a `bool` given that the arguments are already an `opt`, but from what I understand from this choice decreases the chances of something going wrong with the NNS, so fine, let's keep it.

Before merging this we should reflect the changes in the abstract state transition part (unless we've decided to drop that part from the spec).

**Comment by mraszyk** [comment]
> Before merging this we should reflect the changes in the abstract state transition part (unless we've decided to drop that part from the spec).

Good point! I've updated the formal part of the spec to not block this MR on the decision whether we keep it or not.

**Comment by Dfinity-Bjoern** [comment]
Note from the interface spec meeting: Do we have an actual reason for the `opt bool`? @roman-kashitsyn wrote "just in case" but we actually weren't sure whether or not that was a joke.

**Comment by roman-kashitsyn** [comment]
> Note from the interface spec meeting: Do we have an actual reason for the `opt bool`? @roman-kashitsyn wrote "just in case" but we actually weren't sure whether or not that was a joke.

Just in case you want to deprecate this field later.

**Comment by crusso** [comment]
> > Note from the interface spec meeting: Do we have an actual reason for the `opt bool`? @roman-kashitsyn wrote "just in case" but we actually weren't sure whether or not that was a joke.
> 
> Just in case you want to deprecate this field later.

I could be wrong, but I think because it occurs in negative position, as an argument to `install_code`, you could just promote to `reserved` or remove the field from the record anyway. But I always get this wrong.

**Comment by roman-kashitsyn** [comment]
> > > Note from the interface spec meeting: Do we have an actual reason for the `opt bool`? @roman-kashitsyn wrote "just in case" but we actually weren't sure whether or not that was a joke.
> > 
> > 
> > Just in case you want to deprecate this field later.
> 
> I could be wrong, but I think because it occurs in negative position, as an argument to `install_code`, you could just promote to `reserved` or remove the field from the record anyway. But I always get this wrong.

I'm fine either way. The more I use Candid, the more I prefer using optional fields everywhere...

**Comment by crusso** [comment]
Whatever we do about the argument, does type `change_details` in ic.dic needs to evolve too, or is the omission of the `upgrade` payload deliberate? I imagine `change_details` occurs in a positive position, which might force the choice of refining the `upgrade` payload versus adding a new variant (in favour of refining the upgrade payload.).

**Comment by mraszyk** [comment]
> Whatever we do about the argument, does type `change_details` in ic.dic needs to evolve too, or is the omission of the `upgrade` payload deliberate? I imagine `change_details` occurs in a positive position, which might force the choice of refining the `upgrade` payload versus adding a new variant (in favour of refining the upgrade payload.).

Good point! We can either leave `change_details` as they're defined now (meaning we don't expose the fact whether `pre_upgrade` was run, effectively treating it as an implementation detail) or extend the type of the `upgrade` variant in `change_details` to `opt record {...}` which is fine because `null` (the current implicit type of the `upgrade` variant in `change_details`) is a subtype of `opt record {...}`.

**Comment by Dfinity-Bjoern** [comment]
In interface spec meeting: moving from `bool` to `reserved` would also work, so it seems using `bool` instead of `opt bool` seems preferable. We could alternatively use `opt null`, which may be more natural (then it would be `null` or `opt null`).

**Comment by mraszyk** [comment]
> I could be wrong, but I think because it occurs in negative position, as an argument to install_code, you could just promote to reserved or remove the field from the record anyway. But I always get this wrong.

What we missed though is that we need both forwards- and backwards-compatibility when deprecating those fields and thus we need an `opt ...`.

**Review by Dfinity-Bjoern — APPROVED** [review]

**Pull request merged**
